docs/using-the-compiler.rst

```diff
@@ -0,0 +1,249 @@
+find the file there, it will look at ``/usr/local/lib/fallback`` (the empty prefix
+always matches). ``solc`` will not read files from the filesystem that lie outside of
+the remapping targets and outside of the directories where explicitly specified source
+files reside, so things like ``import "/etc/passwd";`` only work if you add ``=/`` as a remapping.
+
+If there are multiple matches due to remappings, the one with the longest common prefix is selected.
+
+If your contracts use :ref:`libraries <libraries>`, you will notice that the bytecode contains substrings of the form ``__LibraryName______``. You can use ``solc`` as a linker meaning that it will insert the library addresses for you at those points:
+
+Either add ``--libraries "Math:0x12345678901234567890 Heap:0xabcdef0123456"`` to your command to provide an address for each library or store the string in a file (one library per line) and run ``solc`` using ``--libraries fileName``.
+
+If ``solc`` is called with the option ``--link``, all input files are interpreted to be unlinked binaries (hex-encoded) in the ``__LibraryName____``-format given above and are linked in-place (if the input is read from stdin, it is written to stdout). All options except ``--libraries`` are ignored (including ``-o``) in this case.
+
+
+**************************************************
+Standardized Input Description and Metadata Output
+**************************************************
+
+In order to ease source code verification of complex contracts that are spread across several files,
+there is a standardized way for describing the relations between those files.
+Furthermore, the compiler can generate a json file while compiling that includes
+the (hash of the) source, natspec comments and other metadata whose hash is included in the
+actual bytecode. Specifically, the creation data for a contract has to begin with
+`push32 <metadata hash> pop`.
+
+The metadata standard is versioned. Future versions are only required to provide the "version" field,
+the "language" field and the two keys inside the "compiler" field.
+The field compiler.keccak should be the keccak hash of a binary of the compiler with the given version.
+
+The example below is presented in a human-readable way. Properly formatted metadata
+should use quotes correctly, reduce whitespace to a minimum and sort the keys of all objects
+to arrive at a unique formatting.
+
+Comments are of course not permitted and used here only for explanatory purposes.
+
+Input Description
+-----------------
+
+QUESTION: How to specific file-reading callback? - probably not as part of json input
+
+The input description is language-specific and could change with each compiler version, but it
+should be backwards compatible if possible.
+
+    {
+      sources:
+      {
+        // the keys here are the "global" names of the source files, imports can use other files via remappings (see below)
```

Comment:
What applies to `sources` in metadata applies here. Two differences:
- `keccak256` is used to verify the content after acquiring it
- `file:` as a url is supported